### ISSUE TITLE
Splitscreen support

### ISSUE THREAD

**Issue by Seamusmario** [GeyserMC/Geyser#750]
**What feature do you want?**
Splitscreen support. I tried to join with splitscreen but the second player is stuck on "Building Terrain".

**Alternatives?**
None.

**Comment by DoctorMacc** [comment]
I'm actually unfamiliar if this is possible, so someone with better experience may come and correct me.
Is the second player also on an Xbox account?

**Comment by DavidTheFrikr** [comment]
The second player is also on an Xbox account AFAIK, I know that Switch split screen players need to have at least a Switch Online membership

**Comment by jackson-57** [comment]
Right, I can confirm the second point in the message above. I don't think the player has to be signed into a Microsoft account, but I might be remembering wrong. I think it used their Nintendo name when connecting.

**Comment by DoctorMacc** [comment]
According to SupremeMortal, this is something that needs to be resolved at the protocol level.

**Comment by mtsr** [comment]
Players on Switch have to be logged into NSO and MS. But second player unfortunately doesn't get his XUID passed on to servers. This is a proper bug since 1.14.60, apparently.

**Comment by jimbt** [comment]
This issue also occurs on PS4 (1.16.100) when connecting to a LAN paper/geyser server.  The server (log) doesn't even recognize that the second player is attempting to connect.  A LAN mojang bedrock server works as expected.

**Comment by maciekhtc** [comment]
I also just tried connecting to my server with geyser running on the laptop inside the same LAN. 
First player joins geyser and server no problem, second player is seeing only loading screen with not moving progress. 
Second player has playstation plus and microsoft account.
The second player can connect to the geyser and server when he tries it before first player, always second player on splitscreen has this problem.
I have tried creating LAN game with Minecraft for Windows10 and both playstation players on splitscreen can connect and play together.
Is this still a work in progress or it might not be resolved?

**Comment by mtsr** [comment]
I did an initial implementation which I haven't updated recently, see #1459 . After some discussion it seemed better to move it to the Protocol library. I haven't gotten to it, but the splitscreen support itself is quite simple as you can see from the PR. The challenge will be in fitting it with the way the Protocol library is designed.

**Comment by alexander7567** [comment]
I spent hours setting up a server and perfecting it for me and my little kids to play together, just to find out split screen isn't supported for xbox :( I second this feature.

**Comment by jimbt** [comment]
> I spent hours setting up a server and perfecting it for me and my little kids to play together, just to find out split screen isn't supported for xbox :( I second this feature.

I resorted to a official bedrock bedrock server (with all its failings) for the very same reason.  The devs are doing awesome work, but this will always be a best effort with unavoidable limitations.  The only complete resolution for full compatibility is bedrock/java feature parity and official support.  Thankfully, this particular issue does have a resolution in sight.

**Comment by mtsr** [comment]
My implementation above worked, but hasn't been updated in a while. It did initially require some changes in other libraries, but I think those PRs were merged. Anyway, since my kids haven't played minecraft in a while, I'm unlikely to update it soon, but feel free to do it, if you can.

**Comment by alexander7567** [comment]
@mtsr thanks - I do see what you are talking about (#976), but I just dont know enough about Java to be able to merge the code and build it to test it. I do wish I could help, but my java is very, very limited. I do hope to see support for this soon, but unfortunately, all I will be able to do is wait I believe. I guess I am stuck with bedrock for now. Thanks for your time in getting the foundation work down!

**Comment by Waffeleisen23** [comment]
Hi there,

first of all, thank you a lot for your great work on Geyser.

Is there any chance that the split screen feature will get implemented? I just have tested it right now and it is stuck in "build world".

Best regards and many thanks

**Comment by lukeeey** [comment]
> Players on Switch have to be logged into NSO and MS. But second player unfortunately doesn't get his XUID passed on to servers. This is a proper bug since 1.14.60, apparently.

i’m pretty sure it’s in the sub client login packet

**Comment by mtsr** [comment]
No, there’s a proper bug in the console client and it doesn’t send a proper ID for the second player. I’m not at my PC, but you can probably find the ticket in the official Minecraft bug tracker if you look for it.

I’m betting there’s an issue with for example switch only having a single active player in general. 

--

Jonas

**Comment by jimbt** [comment]
> No, there’s a proper bug in the console client and it doesn’t send a proper ID for the second player. I’m not at my PC, but you can probably find the ticket in the official Minecraft bug tracker if you look for it. I’m betting there’s an issue with for example switch only having a single active player in general.

For the PS4 client, the second player is always "Steve", but splitscreen otherwise works with the official bedrock server.

**Comment by mtsr** [comment]
It's a client issue, so no, there's nothing that can be done except mojang/microsoft fixing the bedrock client.

**Comment by jimbt** [comment]
Again, the clients work with the official bedrock dedicated server.  The only issue I've seen is that the extra players are always named Steve with default skins.  The extra players do keep all their other data (e.g. location, inventory) so it is playable.

I'll agree the bedrock clients have a bug regarding extra players, but it doesn't break the official servers so why would it be a road block for Geyser?

**Comment by mtsr** [comment]
Geyser can definitely support the same somewhat broken splitscreen that the bedrock standalone server does. In fact, I had my kids playing split screen on a forked version.

Unfortunately it required a few changes in Geyser and its dependencies that would not be merged as is. The current plan is to integrate it into a new version of Protocol, the bedrock protocol library.

**Comment by Morthion** [comment]
@mtsr That's great news and will be hugely welcome. I had the same scenario as above just last night, simply assuming splitscreen would work and then feeling the disappointment shared by myself, my wife and my kiddos for our first family minecraft server (That's not limited by Bedrock). Don't mean to give you a sob story, but simply want to emphasize what this will mean to many families. Thank you for all your hard work.

**Comment by Seamusmario** [comment]
I've been supporting Geyser for 2 years and this issue still gets attention 👍

**Comment by MaPa80711** [comment]
Hello, 

I use the Geyser MC plugin together with Floodgate. 

As in the previous comments, I am also concerned with the splitscreen problem. 

Is there any update on the problem? Will it be developed further? 




Thanks for your support

Greeting MaPa80711

**Comment by markmetcalfe** [comment]
I've updated the implementation done by @mtsr to work with 1.19 in #3109
It uses a workaround to the XUID issue where we just manually record the xbox ID of splitscreen users in the Geyser configuration. Would love a review if you @Camotoy or anyone else has time to.

**Comment by Waffeleisen23** [comment]
> I've updated the implementation done by @mtsr to work with 1.19 in #3109 It uses a workaround to the XUID issue where we just manually record the xbox ID of splitscreen users in the Geyser configuration. Would love a review if you @Camotoy or anyone else has time to.

I am playing on Switch - Would it be helpful when I test it too or is the fix online related to Xbox?

**Comment by mtsr** [comment]
> I've updated the implementation done by @mtsr to work with 1.19 in #3109 It uses a workaround to the XUID issue where we just manually record the xbox ID of splitscreen users in the Geyser configuration. Would love a review if you @Camotoy or anyone else has time to.

Really happy to see this. With the summer holidays coming up, my kids are likely to be playing more minecraft, so this will be awesome to have again.

**Comment by mtsr** [comment]
> I am playing on Switch - Would it be helpful when I test it too or is the fix online related to Xbox? I see that you mention PS so it seems that this is a general solution for consoles?

I think the config file workaround should work fine for Switch too. The console client doesn't send the XUID for the connecting users, but it does still send some info, such as the username, which can then be mapped to a XUID.

**Comment by Camotoy** [comment]
Seeing as we safely have a newer version of this PR, I'll close this one. It seems like 3.0 is taking a longer time than we anticipated, so it'll be nice to have *something* until 3.0 with splitscreen support rolls around. Thanks of course for your contributions.

**Comment by Camotoy** [comment]
I thought this was a PR my bad.

**Comment by MaPa80711** [comment]
Hello all, 

have followed with joy the course here and yesterday, unfortunately without success, the adjustment to the splitscreen support PS4/PS5 tried.

I have proceeded as described in https://github.com/GeyserMC/Geyser/pull/3109 and have extended the config.yml in the Geyser Spigot plugin. 

![grafik](https://user-images.githubusercontent.com/105710637/189479721-07122d42-abc2-4fb2-900f-c0d2b1a86d6b.png)


Unfortunately, remains with me further the 2nd player in the terrain will stand built screen. 

In the log of the Spigot server no messages come. 

Installed is:
minecraft server version 1.19.2
CraftBukkit version 3579-Spigot-e53686f-dd0a218 (MC: 1.19.2) (Implementing API version 1.19.2-R0.1-SNAPSHOT)
floodgate v2.2.0-SNAPSHOT (b73-4f36112)
Multiverse-Core v4.3.1-b861 (API v24)
Geyser-Spigot v2.0.7-SNAPSHOT (git-master-a16fc9c) 

Can someone help me here?

**Comment by Seamusmario** [comment]
> I thought this was a PR my bad.

????

**Comment by MaPa80711** [comment]
Hi @all, are there and updates?

**Comment by Kysaut** [comment]
Do we have an update on this issue? I have tested the old fix to see if it works but as it was stated in another post it was broken. Are we getting anywhere on this or would it be possible to get a download of the old jar to be able to do this again while waiting for an update. I know using older builds is not recommended but for a private server it might not be a bad temporary work around.

**Comment by mseguin** [comment]
Is there any update on this issue ?

**Comment by ImAUs3r** [comment]
Hi, any updates here? Is there a way to support the development?

**Comment by Biofa5000** [comment]
I am on Nintendo Switch using aternos server with spigot 1.21 and have:
-geyser
-flodgate
installed.
Now when I try to go in splitscreen mode on switch it just says: "Invalid packet received!" and both accounts log out of the server.

Both accounts are logged in Nintendo online with membership and also with Microsoft account.

Can the devs fix this? Cause I don't want to wait another 4 years :(

**Comment by onebeastchris** [comment]
Copied from https://github.com/GeyserMC/Geyser/pull/3109#issuecomment-2198225089:

The issue is the bug that has been mentioned here before: https://bugs.mojang.com/browse/MCPE-71033
It's still not fixed, until then it's unlikely that any form of splitscreen support will be added.
